### PR TITLE
RavenDB-10732: [merge from v4.0] generic struct specialization for sm…

### PR DESCRIPTION
**Pull request by aviviadi**
…aps reader

### CODE REVIEW

File: src/Sparrow/LowMemory/LowMemoryNotification.cs [hunk 1]
```diff
@@ -313,8 +314,9 @@ private int CheckMemoryStatus()
 
             var memInfo = MemoryInformation.GetMemoryInfo();
 
-            var currentProcessMemoryMappedShared = GetCurrentProcessMemoryMappedShared();
-            var availableMem = (memInfo.AvailableMemory + currentProcessMemoryMappedShared);
+        var result = new SmapsReader().CalculateMemUsageFromSmaps<SmapsReaderNoAllocResults>();
```
Comment (ayende) [inline src/Sparrow/LowMemory/LowMemoryNotification.cs:317]: Can we avoid this allocation here?

File: src/Sparrow/Platform/Posix/SmapsReader.cs [hunk 2]
```diff
@@ -17,6 +72,7 @@ public class SmapsReader
 
         private const int BufferSize = 4096;
         private readonly byte[][] _smapsBuffer = {new byte[BufferSize], new byte[BufferSize]};
```
Comment (ayende) [inline src/Sparrow/Platform/Posix/SmapsReader.cs:74]: The problem with allocating the `SmapsReader` is that it allocates 8KB at least, so we want to reduce this as much as possible.

File: src/Sparrow/LowMemory/LowMemoryNotification.cs (hunk 1 above)
Comment (aviviadi) [reply src/Sparrow/LowMemory/LowMemoryNotification.cs:317]: we can hold static SmapsReader instance in MemoryInformation.

Comment (ayende) [reply src/Sparrow/LowMemory/LowMemoryNotification.cs:317]: Would that be thread safe, though?

Comment (aviviadi) [reply src/Sparrow/LowMemory/LowMemoryNotification.cs:317]: it is being called from the LowMemoryNotification thread.  We have only one..

Comment (ayende) [reply src/Sparrow/LowMemory/LowMemoryNotification.cs:317]: Then let's allocate that on the method, and pass as argument, rather than a static fifeld.

**Pull request merged**